synapse/storage/controllers/persist_events.py

```diff
@@ -463,6 +479,22 @@ async def persist_event(
         replaced_event = replaced_events.get(event.event_id)
         if replaced_event:
             event = await self.main_store.get_event(replaced_event)
+        else:
+            # We expect events to be persisted by this point and this makes
+            # mypy happy about `stream_ordering` not being optional below
+            assert event.internal_metadata.stream_ordering
+            # Invalidate related caches after we persist a new event
+            relation = relation_from_event(event)
+            self.main_store._invalidate_caches_for_event(
+                stream_ordering=event.internal_metadata.stream_ordering,
+                event_id=event.event_id,
+                room_id=event.room_id,
+                etype=event.type,
+                state_key=event.state_key if hasattr(event, "state_key") else None,
+                redacts=event.redacts,
+                relates_to=relation.parent_id if relation else None,
+                backfilled=backfilled,
+            )
```

Comment:
I originally had the cache invalidation happen in the `_persist_events_txn` layer but that doesn't work because `ValueError: Cache objects can only be accessed from the main thread` and it's already in the background queue thread at that point.

So I moved the cache invalidation to `persist_events` and `persist_event`. Is there potential for things to go wrong if something fails in the middle of the event persist queue where maybe we don't invalidate cache somehow? 🤔

AFAICT, even though `update_current_state` also calls `_event_persist_queue`, it does it's own state updates and cache invalidation. Not sure exactly why that is conflated in this path.

```
persist_events or persist_event or update_current_state (not really)
self._event_persist_queue
_process_event_persist_queue_task
_persist_event_batch
_persist_events_and_state_updates
_persist_events_txn
```

Reply:
It looks like @Fizzadar was able to call `_invalidate_caches_for_event` from `_persist_events_txn`, see https://github.com/matrix-org/synapse/pull/13796/files#diff-abd7b56ac91e8ec195b04b78421459152efe5bd9cba8998e50d23dfb31ebd6c5R426-R436

The `txn.call_after(...)` must be back in the main thread which makes it work there

---

It looks like it would be probably be good to review https://github.com/matrix-org/synapse/pull/13796 first and then I can adapt my PR. If that one lands, I can remove the `_invalidate_caches_for_event` calls here.

But this PR still has a fix for the key errors, https://github.com/matrix-org/synapse/pull/13863#discussion_r977053643

Reply:
Moved `_invalidate_caches_for_event` back `_persist_events_txn` using what [#13796](https://github.com/matrix-org/synapse/pull/13796/files#diff-abd7b56ac91e8ec195b04b78421459152efe5bd9cba8998e50d23dfb31ebd6c5R426-R436) does but that PR should be reviewed first.